### ISSUE TITLE
Certificate realm not working with WebServices in payara 5.194

### ISSUE THREAD

**Issue by borisheithecker** [payara/Payara#4596]
# Description #
----------

I'm upgrading a payara-server-full application from 5.182 to 5.194 
The application is deployed as an ear-file and it has one ejb-jar and two web-jars. 

The ejb-jar defines a webs ervice as follows. I leave out unrelated code. 

```
@WebService(...)
@Stateless
@DeclareRoles({"admin"})
@RolesAllowed({"admin"})
public class .... {
protected SessionContext context;
public xxxx() {
        boolean isAdmin = context.isCallerInRole("admin");
        String name = context.getCallerPrincipal().getName();
}
```

In glassfish-ejb-jar.xml the following is set: 

```
<security-role-mapping>
    <role-name>admin</role-name>
    <group-name>admin</group-name>
 </security-role-mapping>
  <ejb>
      <ejb-name>......</ejb-name>
      <webservice-endpoint>
        <port-component-name>.......</port-component-name>
        <endpoint-address-uri>.....</endpoint-address-uri>
        <login-config>
          <auth-method>CLIENT-CERT</auth-method>
          <realm>admin-cert</realm>
        </login-config>
        <transport-guarantee>CONFIDENTIAL</transport-guarantee>
      </webservice-endpoint>
  </ejb>

```
In the prior version of the app, which was based on 5.182, I could assign the group to the certificate realm by way of defining a custom jaas-context in domain.xml. This is broken since 5.184. I've submitted another related issue https://github.com/payara/Payara/issues/4533 recently. 

To amend this, I've put

```
@CertificateAuthenticationMechanismDefinition
@CertificateIdentityStoreDefinition(value = "admin-cert", assignGroups = "admins")
```

at a suitable place in one of the web-jars in order to assign groups to certificate login. This works in all other cases (other REST services) but not for the SOAP WebService above in the ejb-jar. 

If the `@RolesAllowed({"admin"})` above is present as shown, I get a "Failed permission check". If I comment it out, I get `false` for `isAdmin`, but the right name of the authenticated caller for `name`. 

To summarize, client cert authentication itself works for ejb web services, but group/role assigment is obviously broken. Callers are authenticated but not assigned the provided groups/roles. 

BTW: Setting 
`<property name="default-role-mapping" value="true">`
in glassfish-ejb-jar.xml doesn't make a difference. 

# Expected outcome #
----------

The expected outcome is that the caller must be authenticated and assigned the designated group/role.  `isAdmin` should be `true`. 

If this is a bug, is there a workaround? Is there a switch to turn off all the new authentication features since 5.184 for backward compatibility?

If it's not a bug, please let me know what I'm doing wrong.

**Comment by borisheithecker** [comment]
I could try to contribute, if someone could give me a starting point. Which parts of the sources are responsible for the parsing and processing for the descriptor files (glassfish-ejb-jar.xml, web.xml) and mapping them to JACC?

**Comment by OndroMih** [comment]
HI @borisheithecker, you can start by providing a simple to follow scenario how to reproduce this on the latest version which is now 5.201. A reproducer should ideally follow the SSCCE rules: http://www.sscce.org/. It will greatly help us to find the cause and fix it.

If you'd like to investigate yourself, you can try debugging Payara Server with you application deployed and connecting to the debug port from your IDE, e.g. Netbeans. Most IDEs can deploy your application and debug it and you can debug Payara Server internals in the same way as you debug your application. One way to do it is to download Payara Server source code, build it with `mvn install` and open the maven project of the Payara Server component you want to debug and add it as a provided dependency to your application. Then the IDE will think it's part of your application and will know how to debug the sources.

I would start with adding breakpoints to methods in [CertificateIdentityStoreDefinition.java](https://github.com/payara/Payara/blob/master/appserver/security/realm-stores/src/main/java/fish/payara/security/realm/identitystores/CertificateRealmIdentityStore.java) and [CertificateRealm.java](https://github.com/payara/Payara/blob/master/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/realm/certificate/CertificateRealm.java) and look at what's happening there and which classes call them when you access the WS endpoint.

**Comment by borisheithecker** [comment]
Hi @OndroMih, here are my steps to reproduce the bug: https://github.com/borisheithecker/payara-clientcert-test2
https://github.com/payara/Payara/pull/4610 is my suggested fix for it. I havn't done any testing, and I'm not sure if it breaks any other behaviour. 
The reason for the failure to fully authenticate the caller is that in https://github.com/payara/Payara/blob/master/appserver/security/webintegration/src/main/java/com/sun/web/security/RealmAdapter.java at a call to https://github.com/payara/Payara/blob/master/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/WebAndEjbToJaasBridge.java the provided non-standard realm name is ignored and reset to "certificate". 
My demo code defines a new realm "admin-cert" and assigns groups/roles to it. No groups/roles are assigned to the default realm named "certificate" and therefore the caller is not authenticated as expected. 
I've submitted another issue recently that group assignment is not working if the default realm name is used, see https://github.com/payara/Payara/issues/4533

**Comment by borisheithecker** [comment]
> HI @borisheithecker, you can start by providing a simple to follow scenario how to reproduce this on the latest version which is now 5.201. A reproducer should ideally follow the SSCCE rules: http://www.sscce.org/. It will greatly help us to find the cause and fix it.
> 
> If you'd like to investigate yourself, you can try debugging Payara Server with you application deployed and connecting to the debug port from your IDE, e.g. Netbeans. Most IDEs can deploy your application and debug it and you can debug Payara Server internals in the same way as you debug your application. One way to do it is to download Payara Server source code, build it with `mvn install` and open the maven project of the Payara Server component you want to debug and add it as a provided dependency to your application. Then the IDE will think it's part of your application and will know how to debug the sources.
> 
> I would start with adding breakpoints to methods in [CertificateIdentityStoreDefinition.java](https://github.com/payara/Payara/blob/master/appserver/security/realm-stores/src/main/java/fish/payara/security/realm/identitystores/CertificateRealmIdentityStore.java) and [CertificateRealm.java](https://github.com/payara/Payara/blob/master/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/realm/certificate/CertificateRealm.java) and look at what's happening there and which classes call them when you access the WS endpoint.

**Comment by MeroRai** [comment]
Hi @borisheithecker,

Your PR to fix this issue has now been merged. Thank you very much for your contribution. 

Kind regards,
Mero Rai
